Source/Core/Common/OpenXR.cpp

```diff
@@ -0,0 +1,714 @@
+
+  std::vector<int64_t> swapchain_formats(swapchain_format_count);
+  result = xrEnumerateSwapchainFormats(m_session, swapchain_format_count, &swapchain_format_count,
+                                       swapchain_formats.data());
+
+  if (XR_FAILED(result))
+  {
+    ERROR_LOG(VIDEO, "OpenXR: xrEnumerateSwapchainFormats: %d", result);
+    return false;
+  }
+
+  // Require left/right views have identical sizes
+  if (config_views[0].recommendedImageRectWidth != config_views[1].recommendedImageRectWidth ||
+      config_views[0].recommendedImageRectHeight != config_views[1].recommendedImageRectHeight ||
+      config_views[0].recommendedSwapchainSampleCount !=
+          config_views[1].recommendedSwapchainSampleCount)
+  {
+    ERROR_LOG(VIDEO, "OpenXR: Recommended left/right swapchain image sizes differ.");
+    return false;
+  }
+
+  auto& selected_config_view = config_views[0];
+  m_swapchain_size = {int32_t(selected_config_view.recommendedImageRectWidth),
+                      int32_t(selected_config_view.recommendedImageRectHeight)};
+
+  const auto format = std::find_first_of(swapchain_formats.begin(), swapchain_formats.end(),
+                                         supported_formats.begin(), supported_formats.end());
+  if (format == swapchain_formats.end())
+  {
+    ERROR_LOG(VIDEO, "OpenXR: No usable swapchain formats.");
+    return false;
+  }
+
+  const auto selected_swapchain_format = *format;
+  m_swapchain_format = selected_swapchain_format;
+
+  DEBUG_LOG(VIDEO, "OpenXR: Using swapchain format: %d", m_swapchain_format);
+
+  // Create swapchain with texture array of size view_count.
+  XrSwapchainCreateInfo swapchain_info{XR_TYPE_SWAPCHAIN_CREATE_INFO};
+  swapchain_info.createFlags = 0;
+  // TODO: Are these flags sane?
+  swapchain_info.usageFlags =
+      XR_SWAPCHAIN_USAGE_SAMPLED_BIT | XR_SWAPCHAIN_USAGE_COLOR_ATTACHMENT_BIT;
+  swapchain_info.format = selected_swapchain_format;
+  swapchain_info.sampleCount = selected_config_view.recommendedSwapchainSampleCount;
+  swapchain_info.width = selected_config_view.recommendedImageRectWidth;
+  swapchain_info.height = selected_config_view.recommendedImageRectHeight;
+  swapchain_info.faceCount = 1;
+  swapchain_info.arraySize = view_count;
+  swapchain_info.mipCount = 1;
+
+  result = xrCreateSwapchain(m_session, &swapchain_info, &m_swapchain);
+  if (XR_FAILED(result))
+  {
+    ERROR_LOG(VIDEO, "OpenXR: xrCreateSwapchain: %d", result);
+    return false;
+  }
+
+  INFO_LOG(VIDEO, "OpenXR: Created swapchain: %d x %d (%d samples)", swapchain_info.width,
+           swapchain_info.height, swapchain_info.sampleCount);
+
+  return true;
+}
+
+bool Session::EnumerateSwapchainImages(uint32_t count, uint32_t* capacity, void* data)
+{
+  const XrResult result = xrEnumerateSwapchainImages(
+      m_swapchain, count, capacity, static_cast<XrSwapchainImageBaseHeader*>(data));
+  return XR_SUCCEEDED(result);
+}
+
+// TODO: Allow for a timeout? return optional<> ?
```

Comment:
Not sure if this was intended, but the method body looks as if you wrote it with `std::optional` in mind; at least I don't think anyone would consciously use `{}` instead of `0` (or any other magic number for that matter) as return value for an `uint32_t`. The only thing missing is the actual return type.